---
title: "R Notebook"
output: html_notebook
---

Table 1
```{r}
library(RSQLite)
library(SRAdb)
library(tidyverse)
library(stringr)
# getSRAdbFile(destdir='/Volumes/ThunderBay/PROJECTS/mcgaughey/unified_gene_expression/',destfile='SRAmetadb.sqlite.gz')
sqlfile <- '/Volumes/ThunderBay/PROJECTS/mcgaughey/unified_gene_expression/SRAmetadb_2017-01-19.sqlite'
sra_con <- dbConnect(RSQLite::SQLite(),sqlfile)
human_transcriptome_sra_info <- function(keyword) {
  sql_query <- 
    'select *  from sra WHERE library_source="TRANSCRIPTOMIC" AND 
  (study_abstract LIKE keyword OR
  experiment_name LIKE keyword OR
  study_name LIKE keyword OR 
  sample_ID LIKE keyword OR
  sample_name LIKE keyword OR
  study_title LIKE keyword OR
  study_description LIKE keyword OR
  description LIKE keyword) AND 
  library_strategy!="miRNA-Seq" AND
  taxon_id="9606"'
  new_query <- gsub('keyword',keyword, sql_query)
  dbGetQuery(sra_con, new_query)
}
human_tx_studies <- rbind(
  human_transcriptome_sra_info("\"%RPE%\""),
  human_transcriptome_sra_info("\"%macula%\""),
  human_transcriptome_sra_info("\"%fovea%\""),
  human_transcriptome_sra_info("\"%retina%\""),
  human_transcriptome_sra_info("\"%choroid%\""),
  human_transcriptome_sra_info("\"%sclera%\""),
  human_transcriptome_sra_info("\"%iris%\""),
  human_transcriptome_sra_info("\"%lens%\""),
  human_transcriptome_sra_info("\"%cornea%\""),
  human_transcriptome_sra_info("\"%eye%\"")) %>% distinct()

load('~/git/unified_gene_expression/data/eye_studies_considered_2017-01-19.Rdata')
source('~/git/unified_gene_expression/scripts/parse_sample_attribute.R')
source('~/git/scripts/theme_Publication.R')
```

```{r}
human_tx_studies %>% group_by(study_accession) %>% summarise(count=n())
human_tx_studies %>% group_by(sample_accession) %>% summarise(count=n())

core_eye_info %>% group_by(study_accession) %>% summarise(count=n()) %>% arrange(-count)
core_eye_info %>% group_by(sample_accession) %>% summarise(count=n())

core_eye_info %>% group_by(Tissue) %>% summarise(count=n()) %>% arrange(-count)
```

Stats pre-QC filtering
```{r}
core_eye_info %>% 
  mutate(`Study Accession`=ifelse(grepl(x = study_accession, pattern = 'E_MTAB_4377'),'E-MTAB-4377', study_accession)) %>% 
  group_by(`Study Accession`, Tissue, Origin) %>% 
  summarise(`Sample Count`=n()) %>% 
  ggplot(.,aes(y=`Sample Count`, x=`Study Accession`,fill=Tissue)) + 
    geom_bar(stat='identity',position='dodge') +
    theme_Publication() +
    theme(axis.text.x=element_text(angle = 90, hjust = 0))

core_eye_info %>% 
  group_by(Tissue, Origin) %>% 
  summarise(`Sample Count`=n()) %>% 
  ggplot(.,aes(y=`Sample Count`, x=Tissue,fill=Origin)) + 
    geom_bar(stat='identity') + 
    facet_wrap(~Origin) + 
    theme_Publication() +
    theme(axis.text.x=element_text(angle = 90, hjust = 0))
```

Stats post-QC filtering. Samples removed for two reasons: low median counts (which lead from low sequencing depth and/or poor alignment rate) and mis-matched clustering (tissue mis-labeling or errors in acquisition).
```{r}
load('~/git/unified_gene_expression/data/lengthScaledTPM_processed_2017_02.Rdata')
library(superheat)
core_tight <- core_tight %>% dplyr::select(-run_accession)
core_tight <- core_tight[!duplicated(core_tight),]
core_tight$Tissue = trimws(core_tight$Tissue)
core_eye_info_QC <- core_tight %>% 
  filter(sample_accession %in% colnames(lengthScaledTPM_processed)) %>% 
  filter(!Tissue %in% 'ENCODE Cell Line') %>% 
  filter(study_accession!='SRP012682') %>% 
  filter(!sample_accession %in% c('SRS523795','SRS360124','SRS360123'))


core_eye_info_QC %>% 
  mutate(`Study Accession`=ifelse(grepl(x = study_accession, pattern = 'E_MTAB_4377'),'E-MTAB-4377', study_accession)) %>% 
  group_by(`Study Accession`, Tissue, Origin) %>% 
  summarise(`Sample Count`=n()) %>% 
  ggplot(.,aes(y=`Sample Count`, x=`Study Accession`,fill=Tissue)) + 
    geom_bar(stat='identity',position='dodge') +
    theme_Publication() +
    theme(axis.text.x=element_text(angle = 90, hjust = 0))
```

```{r, fig.width=2,fig.height=1.5}
#heatmap of tissue vs origin
heatmap_eye <- core_eye_info_QC %>% 
  group_by(Tissue, Origin) %>% 
  summarise(Count=n()) %>% 
  spread(Tissue,Count,fill=0) %>% 
  data.frame()
row_order <- c('Cell Line', 'Stem Cell Line', 'Fetal Tissue', 'Adult Tissue')
heatmap_eye <- heatmap_eye[match(row_order, heatmap_eye$Origin),]
row.names(heatmap_eye) <- heatmap_eye$Origin
heatmap_eye <- heatmap_eye[,2:5]
heatmap_eye <- heatmap_eye %>% select(Cornea, Retina, RPE, ESC)
superheat(heatmap_eye, 
          heat.col.scheme = 'purple', 
          X.text = as.matrix(heatmap_eye),
          bottom.label.text.angle = 90,
          bottom.label.size = 0.6,
          left.label.size = 1.0,
          pretty.order.rows = FALSE,
          legend = FALSE,
          grid.hline.col = 'white',
          grid.vline.col = 'white')
```
