---
title: "DE"
output:
  html_notebook: default
  html_document: default
---

We begin by loading in the the raw data, excluding files which fail QC due to either 

```{r, message = FALSE}
library(tximport)
library(limma)
library(edgeR)
library(qsmooth)

library(superheat)
source('~/git/scripts/theme_Publication.R')
source('~/git/unified_gene_expression/scripts/parse_sample_attribute.R')
# low median count files removed, from 'analysis/QC_and_qsmooth.Rmd'
load('~/git/unified_gene_expression/data/lengthScaledTPM_processed_2017_02.Rdata')
```

Clean up metadata, removing run accession, and then removing dups. Some samples have multiple runs
```{r, message = FALSE}
core_tight <- core_tight %>% dplyr::select(-run_accession)
core_tight <- core_tight[!duplicated(core_tight),]
core_tight$Tissue = trimws(core_tight$Tissue)
```

Pull in salmon counts data directories
```{r, message = FALSE}
# biowulf Salmon counts 
working_dir <- '/data/mcgaugheyd/projects/nei/mcgaughey/unified_gene_expression/salmon_counts_bootstrap50_txUsed/'
# eyeMac
working_dir <- '/Volumes/ThunderBay/PROJECTS/mcgaughey/unified_gene_expression/salmon_counts_bootstrap50_txUsed/'
setwd(working_dir)
files <- list.files(path=working_dir,recursive=TRUE,pattern='quant.sf')
```

Eye only samples, with outliers removed, from previous analyses
```{r, message = FALSE}
eye_samples <- core_tight %>% 
  filter(sample_accession %in% colnames(lengthScaledTPM_processed)) %>%  # low median count outliers
  filter(!Tissue %in% 'ENCODE Cell Line') %>% 
  filter(study_accession!='SRP012682') %>% 
  filter(!sample_accession %in% c('SRS523795','SRS360124','SRS360123')) %>%  # retina in RPE
  .[['sample_accession']] 
```

Now remove outlier files and ENCODE files. Then select representative subset of GTEx by matching number of tissues to eye tissues and taking equal representation across the 22 GTEx tissues
```{r, message = FALSE} 
set.seed(123421)
sample_number <- length(eye_samples) /
  (core_tight %>% filter(sample_accession %in% colnames(lengthScaledTPM_processed)) %>% filter(study_accession=='SRP012682') %>% dplyr::select(Tissue) %>% unique() %>% nrow())
sample_number <- round(sample_number)
gtex_sub_samples <- core_tight %>% 
  filter(sample_accession %in% colnames(lengthScaledTPM_processed)) %>% # low median count outliers
  filter(!Tissue %in% 'ENCODE Cell Line') %>% 
  filter(!sample_accession %in% eye_samples) %>% 
  filter(!sample_accession %in% c('SRS523795','SRS360124','SRS360123')) %>%  # retina in RPE
  filter(!sample_accession %in% c('SRS332999', 'SRS389531', 'SRS626623', 'SRS627133', 'SRS623923', 'SRS629483')) %>% # few gtex samples that outliers (in weird clusters)
  group_by(Tissue) %>% 
  sample_n(sample_number) %>% 
  .[['sample_accession']]
eye_and_gtex_samples <- c(eye_samples, gtex_sub_samples)
```

Update files to import and get eye_(and_gtex)_samples in same order as eye_(and_gtex)_files
```{r, message = FALSE}
eye_and_gtex_files <- files[files %in% paste(eye_and_gtex_samples,'/quant.sf',sep='')]
eye_files <- files[files %in% paste(eye_samples,'/quant.sf',sep='')]
eye_samples <- gsub('\\/quant.sf','',eye_files)
eye_and_gtex_samples <- gsub('\\/quant.sf','',eye_and_gtex_files)
```
Gene TX to name conversion annotations
```{r, message = FALSE}
load('~/git/unified_gene_expression/data/gencode_v25_annotation.Rdata')
anno <- gencode_v25_annotation %>% dplyr::select(Transcript.ID, Gene.Name)
```

Create metadata file for eye samples
```{r, message = FALSE}
# tximport to aggregate counts to gene level
#txi.eye_and_gtex <- tximport(eye_and_gtex_files, type = "salmon", tx2gene = anno, reader = read_tsv)
#txi.eye <- tximport(eye_files, type = "salmon", tx2gene = anno, reader = read_tsv)

# metadata
SampleTable_eye <- data.frame(eye_samples) %>% 
  mutate(sample_accession = eye_samples) %>% 
  select(sample_accession) %>% 
  left_join(.,core_tight) %>% 
  select(sample_accession, Tissue, Origin)
```

Create metadata file for eye and gtex samples
```{r, message = FALSE}
# metadata
SampleTable_eye_and_gtex <- data.frame(eye_and_gtex_samples) %>% 
  mutate(sample_accession = eye_and_gtex_samples) %>% 
  select(sample_accession) %>% 
  left_join(.,core_tight) %>% 
  select(sample_accession, Tissue, Origin)
```


DESeq2
```{r, message = FALSE}
#############
#DESeq TOO DAMN SLOW
# DESeq object creation for DE
# dds_eye <- DESeqDataSetFromTximport(txi.eye, SampleTable, ~Tissue + Origin)

# differential expression analysis
#DESeq2Table <- DESeq(dds_eye)
#DESeq2Res <- results(DESeq2Table, filterFun=ihw, addMLE=TRUE)
#print(head(DESeq2Res))
#print(table(DESeq2Res$padj < 0.05))
###############
```

tximport for limma voom (eye)
```{r, message = FALSE}
setwd(working_dir)
## eye only
txi.eye.lsTPM_full <- tximport(eye_files, type = "salmon", tx2gene = anno, reader = read_tsv, countsFromAbundance = "lengthScaledTPM")
txi.eye.lsTPM <- data.frame(txi.eye.lsTPM_full$counts)
colnames(txi.eye.lsTPM) <- gsub('\\/quant.sf','',eye_files)
```
tximport for limma voom (eye and gtex)
```{r, message = FALSE}
setwd(working_dir)
## eye and gtex
txi.eyeandgtex.lsTPM_full <- tximport(eye_and_gtex_files, type = "salmon", tx2gene = anno, reader = read_tsv, countsFromAbundance = "lengthScaledTPM")
txi.eye.and.gtex.lsTPM <- data.frame(txi.eyeandgtex.lsTPM_full$counts)
colnames(txi.eye.and.gtex.lsTPM) <- eye_and_gtex_samples
```

Function to do some QC steps and make density plots before and after doing:
1. Remove low count genes
2. scale for library sizes
3. qsmooth
Returns qsmooth'ed set with low count genes removed and scaled for library sizes
```{r, message = FALSE}

QC <- function(data, metadata, qsmoothFactor){
  # density plot, pre low gene count removal and qsmooth(?)
  gather_lST<-gather(data, sample_accession) %>% left_join(.,metadata)
  print(ggplot(gather_lST,aes(x=log2(value+1),group=sample_accession)) +
          geom_density() +
          facet_wrap(~Tissue) + 
          ggtitle('No normalization') +
          theme_Publication())
  
  # remove genes with an average lsTPM < 1
  table(rowSums(data)<(ncol(data)))
  data_geneRemoval <- data[(rowSums(data)>ncol(data)),]
  
  # re-do density plot
  gather_lST<-gather(data_geneRemoval,sample_accession) %>% left_join(.,metadata)
  print(ggplot(gather_lST,aes(x=log2(value+1),group=sample_accession)) +
          geom_density() +
          facet_wrap(~Tissue) +
          ggtitle('Low gene count removal') +
          theme_Publication())
  
  # library size normalize
  norm <- DGEList(data_geneRemoval)
  norm <- calcNormFactors(norm)
  tpm <- norm$counts
  correction <- norm$samples %>% data.frame() %>% .[['norm.factors']]
  lsTPM_librarySize <- tpm %*% diag(correction)
  colnames(lsTPM_librarySize) <- colnames(data_geneRemoval)
  
  # re-redo density plot
  gather_lST<-gather(data.table(lsTPM_librarySize),sample_accession) %>% left_join(.,metadata)
  print(ggplot(gather_lST,aes(x=log2(value+1),group=sample_accession)) +
          geom_density() +
          facet_wrap(~Tissue) +
          ggtitle('Low gene count removal AND library size normalize') +
          theme_Publication())
  
  # qsmooth
  qs <- qsmooth(object = lsTPM_librarySize,groupFactor = as.factor(metadata[,qsmoothFactor]))
  lsTPM_qsmooth <- qsmoothData(qs)
  
  # re-re-redo density plot
  gather_lST<-gather(data.table(lsTPM_qsmooth),sample_accession) %>% left_join(.,metadata)
  print(ggplot(gather_lST,aes(x=log2(value+1),group=sample_accession)) +
          geom_density() +
          facet_wrap(~Tissue) +
          ggtitle('Low gene count removal AND library size normalize AND qsmooth') +
          theme_Publication())

  
  return(lsTPM_qsmooth)
}
```

QC eye samples
```{r, message = FALSE}
lsTPM_qsmooth <- QC(txi.eye.lsTPM, SampleTable_eye, 'Tissue')
```

QC eye and gtex samples
```{r, message = FALSE}
lsTPM_eye_and_gtex_qsmooth <- QC(txi.eye.and.gtex.lsTPM, SampleTable_eye_and_gtex, 'Tissue')
```

Set up design matrix for eye only tests and do voom normalization
```{r, message = FALSE}
# design matrix
tissue_origin <- paste(SampleTable_eye$Tissue, SampleTable_eye$Origin,sep="_") %>% str_replace_all(., " ", ".") %>% factor()
tissue_origin %>% sort() %>% unique()
design <- model.matrix(~0 + tissue_origin)
colnames(design) <- levels(tissue_origin)
# bring in lsTPM valuesvoom
y <- DGEList(lsTPM_qsmooth)
y <- calcNormFactors(y)
v <- voom(y, design)
vfit <- lmFit(v, design)

# tiny function to update contrasts, return topTableF results for multi-comparison
# one-way anova, essentially
limma_voom_test <- function(lmFit_object, your_custom_contrast_matrix){
  vfit.cont <- contrasts.fit(lmFit_object, contrasts = your_custom_contrast_matrix)
  efit <- eBayes(vfit.cont)
  print(summary(decideTests(efit, lfc = 1, p.value = 0.05, adjust.method = 'fdr')))
  efit
  #topTableF(efit, number=30000, adjust.method='fdr')
}
```

Set up design matrix for eye and gtex tests and do voom normalization
```{r, message = FALSE}
# design matrix
new_tissue <- SampleTable_eye_and_gtex %>% mutate(Tissue = ifelse(Tissue %in% c('Retina','RPE','Cornea','ESC'), Tissue, 'Body')) %>% .[['Tissue']]
tissue_origin <- paste(new_tissue, SampleTable_eye_and_gtex$Origin,sep="_") %>% str_replace_all(., " ", ".") %>% factor()
tissue_origin %>% sort() %>% unique()
design_eye_and_gtex <- model.matrix(~0 + tissue_origin)
colnames(design_eye_and_gtex) <- levels(tissue_origin)
# bring in lsTPM valuesvoom
y_eye_gtex <- DGEList(lsTPM_eye_and_gtex_qsmooth)
y_eye_gtex <- calcNormFactors(y_eye_gtex)
v_eye_gtex <- voom(y_eye_gtex, design_eye_and_gtex)
vfit_eye_gtex <- lmFit(v_eye_gtex, design_eye_and_gtex)

```

Six DE tests for eye only set:
1. Cornea vs Retina vs RPE (all adult)
2. ESC vs RPE vs Retina (all Stem cell lines, non immortalized)
3. Adult RPE vs Fetal RPE vs Cell Line RPE
4. immortalized RPE vs Stem Cell Line RPE
5. Adult Retina vs Stem Cell Line Retina
6. Adult Cornea vs Fetal Cornea vs Cell Line Cornea

Test 1 (Cornea vs Retina vs RPE (all adult))
```{r, message = FALSE}
cont.matrix <- makeContrasts(Cornea.adult_vs_Retina.adult='Cornea_Adult.Tissue-Retina_Adult.Tissue',
                             Cornea.adult_vs_RPE.adult='Cornea_Adult.Tissue-RPE_Adult.Tissue',
                             Retina.adult_vs_RPE.adult='Retina_Adult.Tissue-RPE_Adult.Tissue', 
                             levels=design)
test1 <- limma_voom_test(vfit, cont.matrix)
```

Test 2 (ESC vs RPE vs Retina (all cell lines, non immortalized))
```{r, message = FALSE}
cont.matrix <- makeContrasts(ESC.stemcell_vs_Retina.stemcell='ESC_Stem.Cell.Line-Retina_Stem.Cell.Line',
                             ESC.stemcell_vs_RPE.stemcell='ESC_Stem.Cell.Line-RPE_Stem.Cell.Line',
                             Retina.stemcell_vs_RPE.stemcell='Retina_Stem.Cell.Line-RPE_Stem.Cell.Line', 
                             levels=design) 
test2 <- limma_voom_test(vfit, cont.matrix)
```

Test 3 (Adult RPE vs Fetal RPE vs Stem Cell Line RPE)
```{r, message = FALSE}
cont.matrix <- makeContrasts(RPE.stemcell_vs_RPE.fetal='RPE_Stem.Cell.Line-RPE_Fetal.Tissue',
                             RPE.stemcell_vs_RPE.adult='RPE_Stem.Cell.Line-RPE_Adult.Tissue',
                             RPE.fetal_vs_RPE.adult='RPE_Fetal.Tissue-RPE_Adult.Tissue', 
                             levels=design) 
test3 <- limma_voom_test(vfit, cont.matrix)
```

Test 4 (Stem Cell Line RPE vs immortalized cell line RPE)
```{r, message = FALSE}
cont.matrix <- makeContrasts(RPE.stemcell_vs_RPE.immortalizedcell="RPE_Stem.Cell.Line-RPE_Cell.Line",
                             levels=design) 
test4 <- limma_voom_test(vfit, cont.matrix)
```

Test 5 (Adult Retina vs Stem Cell Line Retina)
```{r, message = FALSE}
cont.matrix <- makeContrasts(Retina.adult_vs_Retina.stemcell="Retina_Adult.Tissue-Retina_Stem.Cell.Line",
                             levels=design) 
test5 <- limma_voom_test(vfit, cont.matrix)
```

Test 6 (Adult Cornea vs Fetal Cornea vs Cell Line Cornea)
```{r, message = FALSE}
cont.matrix <- makeContrasts(Cornea.adult_vs_Cornea.fetal="Cornea_Adult.Tissue-Cornea_Fetal.Tissue",
                             Cornea.adult_vs_Cornea.immortalizedcell='Cornea_Adult.Tissue-Cornea_Cell.Line',
                             Cornea.fetal_vs_Cornea.immortalizedcell='Cornea_Fetal.Tissue-Cornea_Cell.Line',
                             levels=design) 
test6 <- limma_voom_test(vfit, cont.matrix)
```

Save 6 tests in one file
```{r, message = FALSE}
big_six <- list(test1,test2,test3,test4,test5,test6)
save(big_six, file='~/git/unified_gene_expression/data/big_six_DE_correct.Rdata')
```

##############################################
Make all pairwise comparisons for eye and gtex
##############################################
```{r, message=FALSE}
combn(unique(tissue_origin),2) %>% 
  t() %>% 
  data.table() %>% 
  mutate(name=paste(V2,V1,sep='_vs_'),
         contrast=paste(V2,V1,sep='-'),
         all_makeContrasts=paste(name,'=\"',contrast,'\",\n',sep='')) %>% 
  .[['all_makeContrasts']] %>% cat()
```

Now make contrast with all comparisons, copied from chunk above.
```{r, message=F}

cont.matrix_all <- makeContrasts(Retina_Adult.Tissue_vs_Body_Tissue="Retina_Adult.Tissue-Body_Tissue",
 RPE_Fetal.Tissue_vs_Retina_Adult.Tissue="RPE_Fetal.Tissue-Retina_Adult.Tissue",
 RPE_Stem.Cell.Line_vs_Retina_Adult.Tissue="RPE_Stem.Cell.Line-Retina_Adult.Tissue",
 RPE_Cell.Line_vs_Retina_Adult.Tissue="RPE_Cell.Line-Retina_Adult.Tissue",
 ESC_Stem.Cell.Line_vs_Retina_Adult.Tissue="ESC_Stem.Cell.Line-Retina_Adult.Tissue",
 RPE_Adult.Tissue_vs_Retina_Adult.Tissue="RPE_Adult.Tissue-Retina_Adult.Tissue",
 Cornea_Adult.Tissue_vs_Retina_Adult.Tissue="Cornea_Adult.Tissue-Retina_Adult.Tissue",
 Cornea_Fetal.Tissue_vs_Retina_Adult.Tissue="Cornea_Fetal.Tissue-Retina_Adult.Tissue",
 Cornea_Cell.Line_vs_Retina_Adult.Tissue="Cornea_Cell.Line-Retina_Adult.Tissue",
 Retina_Stem.Cell.Line_vs_Retina_Adult.Tissue="Retina_Stem.Cell.Line-Retina_Adult.Tissue",
 RPE_Fetal.Tissue_vs_Body_Tissue="RPE_Fetal.Tissue-Body_Tissue",
 RPE_Stem.Cell.Line_vs_Body_Tissue="RPE_Stem.Cell.Line-Body_Tissue",
 RPE_Cell.Line_vs_Body_Tissue="RPE_Cell.Line-Body_Tissue",
 ESC_Stem.Cell.Line_vs_Body_Tissue="ESC_Stem.Cell.Line-Body_Tissue",
 RPE_Adult.Tissue_vs_Body_Tissue="RPE_Adult.Tissue-Body_Tissue",
 Cornea_Adult.Tissue_vs_Body_Tissue="Cornea_Adult.Tissue-Body_Tissue",
 Cornea_Fetal.Tissue_vs_Body_Tissue="Cornea_Fetal.Tissue-Body_Tissue",
 Cornea_Cell.Line_vs_Body_Tissue="Cornea_Cell.Line-Body_Tissue",
 Retina_Stem.Cell.Line_vs_Body_Tissue="Retina_Stem.Cell.Line-Body_Tissue",
 RPE_Stem.Cell.Line_vs_RPE_Fetal.Tissue="RPE_Stem.Cell.Line-RPE_Fetal.Tissue",
 RPE_Cell.Line_vs_RPE_Fetal.Tissue="RPE_Cell.Line-RPE_Fetal.Tissue",
 ESC_Stem.Cell.Line_vs_RPE_Fetal.Tissue="ESC_Stem.Cell.Line-RPE_Fetal.Tissue",
 RPE_Adult.Tissue_vs_RPE_Fetal.Tissue="RPE_Adult.Tissue-RPE_Fetal.Tissue",
 Cornea_Adult.Tissue_vs_RPE_Fetal.Tissue="Cornea_Adult.Tissue-RPE_Fetal.Tissue",
 Cornea_Fetal.Tissue_vs_RPE_Fetal.Tissue="Cornea_Fetal.Tissue-RPE_Fetal.Tissue",
 Cornea_Cell.Line_vs_RPE_Fetal.Tissue="Cornea_Cell.Line-RPE_Fetal.Tissue",
 Retina_Stem.Cell.Line_vs_RPE_Fetal.Tissue="Retina_Stem.Cell.Line-RPE_Fetal.Tissue",
 RPE_Cell.Line_vs_RPE_Stem.Cell.Line="RPE_Cell.Line-RPE_Stem.Cell.Line",
 ESC_Stem.Cell.Line_vs_RPE_Stem.Cell.Line="ESC_Stem.Cell.Line-RPE_Stem.Cell.Line",
 RPE_Adult.Tissue_vs_RPE_Stem.Cell.Line="RPE_Adult.Tissue-RPE_Stem.Cell.Line",
 Cornea_Adult.Tissue_vs_RPE_Stem.Cell.Line="Cornea_Adult.Tissue-RPE_Stem.Cell.Line",
 Cornea_Fetal.Tissue_vs_RPE_Stem.Cell.Line="Cornea_Fetal.Tissue-RPE_Stem.Cell.Line",
 Cornea_Cell.Line_vs_RPE_Stem.Cell.Line="Cornea_Cell.Line-RPE_Stem.Cell.Line",
 Retina_Stem.Cell.Line_vs_RPE_Stem.Cell.Line="Retina_Stem.Cell.Line-RPE_Stem.Cell.Line",
 ESC_Stem.Cell.Line_vs_RPE_Cell.Line="ESC_Stem.Cell.Line-RPE_Cell.Line",
 RPE_Adult.Tissue_vs_RPE_Cell.Line="RPE_Adult.Tissue-RPE_Cell.Line",
 Cornea_Adult.Tissue_vs_RPE_Cell.Line="Cornea_Adult.Tissue-RPE_Cell.Line",
 Cornea_Fetal.Tissue_vs_RPE_Cell.Line="Cornea_Fetal.Tissue-RPE_Cell.Line",
 Cornea_Cell.Line_vs_RPE_Cell.Line="Cornea_Cell.Line-RPE_Cell.Line",
 Retina_Stem.Cell.Line_vs_RPE_Cell.Line="Retina_Stem.Cell.Line-RPE_Cell.Line",
 RPE_Adult.Tissue_vs_ESC_Stem.Cell.Line="RPE_Adult.Tissue-ESC_Stem.Cell.Line",
 Cornea_Adult.Tissue_vs_ESC_Stem.Cell.Line="Cornea_Adult.Tissue-ESC_Stem.Cell.Line",
 Cornea_Fetal.Tissue_vs_ESC_Stem.Cell.Line="Cornea_Fetal.Tissue-ESC_Stem.Cell.Line",
 Cornea_Cell.Line_vs_ESC_Stem.Cell.Line="Cornea_Cell.Line-ESC_Stem.Cell.Line",
 Retina_Stem.Cell.Line_vs_ESC_Stem.Cell.Line="Retina_Stem.Cell.Line-ESC_Stem.Cell.Line",
 Cornea_Adult.Tissue_vs_RPE_Adult.Tissue="Cornea_Adult.Tissue-RPE_Adult.Tissue",
 Cornea_Fetal.Tissue_vs_RPE_Adult.Tissue="Cornea_Fetal.Tissue-RPE_Adult.Tissue",
 Cornea_Cell.Line_vs_RPE_Adult.Tissue="Cornea_Cell.Line-RPE_Adult.Tissue",
 Retina_Stem.Cell.Line_vs_RPE_Adult.Tissue="Retina_Stem.Cell.Line-RPE_Adult.Tissue",
 Cornea_Fetal.Tissue_vs_Cornea_Adult.Tissue="Cornea_Fetal.Tissue-Cornea_Adult.Tissue",
 Cornea_Cell.Line_vs_Cornea_Adult.Tissue="Cornea_Cell.Line-Cornea_Adult.Tissue",
 Retina_Stem.Cell.Line_vs_Cornea_Adult.Tissue="Retina_Stem.Cell.Line-Cornea_Adult.Tissue",
 Cornea_Cell.Line_vs_Cornea_Fetal.Tissue="Cornea_Cell.Line-Cornea_Fetal.Tissue",
 Retina_Stem.Cell.Line_vs_Cornea_Fetal.Tissue="Retina_Stem.Cell.Line-Cornea_Fetal.Tissue",
 Retina_Stem.Cell.Line_vs_Cornea_Cell.Line="Retina_Stem.Cell.Line-Cornea_Cell.Line", levels=design_eye_and_gtex)
```

Redo fit with new contrasts matrix and get F statistic for ENTIRE group. Finding genes where are significantly DE in across all comparisons. Also requiring logFC to be >2 for at least one comparison. 
```{r, message = FALSE}
vfit_all <- lmFit(v_eye_gtex, design_eye_and_gtex)
vfit_all <- contrasts.fit(vfit_all, contrasts=cont.matrix_all)
efit_all <- eBayes(vfit_all)
DE_55 <- topTableF(efit_all,number=300000, adjust.method = 'fdr') %>% data.frame()
save(efit_all, file='~/git/unified_gene_expression/data/limma_voom_DE_all_by_all.Rdata')
```


######################################
Ignore below for now
######################################

Now kmeans cluster the DE genes into groups. First use scree plot to pick n.
```{r, message = FALSE}
# set.seed(3939)
# wss <- (nrow(DE[,1:36])-1)*sum(apply(DE[,1:36],2,var))
# for (i in 2:80) wss[i] <- sum(kmeans(DE[,1:36],
#   	centers=i, iter.max = 10000)$withinss)
# plot(1:80, wss, type="b", xlab="Number of Clusters",
#   ylab="Within groups sum of squares")
```

Do kmeans with 20 groups
```{r, message = FALSE}
# set.seed(9942389)
# DE <- topTableF(efit_all,number=300000, adjust.method = 'fdr', lfc = 2) %>% rownames_to_column('Gene') %>% filter(adj.P.Val<0.01) %>% data.frame()
# kfit <- kmeans(DE[,2:37], 20, iter.max = 10000) # 20 cluster solution
# # move k means cluster assignment over to DE data frame
# DE$cluster <- kfit$cluster
# # clean up the contrast names
# comparisons <- colnames(DE)[2:37] %>% 
#   gsub('_vs_','|vs|', .) %>% 
#   gsub('_',' (',.) %>% 
#   gsub('\\.Line|\\.Tissue',')',.) %>% 
#   gsub('\\|vs\\|',' vs ',.) %>% 
#   gsub('hTERT\\.Cell','hTERT',.)
# colnames(DE)[2:37] <- comparisons
```

```{r, message = FALSE, fig.width=3, fig.height=3}
# for (i in 1:20){
#   cluster_num = i
#   comparisons_to_highlight <- DE %>% 
#     filter(cluster==cluster_num) %>% 
#     select(1:36) %>% 
#     summarise_each(funs(mean)) %>% t() %>% data.frame() %>% 
#     rownames_to_column('Comparison')  %>% 
#     `colnames<-`(c('Comparison', 'Mean')) %>% 
#     mutate(Tercile=ntile(Mean,n=3), Size=0) %>% 
#     mutate(Points=ifelse((abs(Mean)>2 & Tercile %in% c(1,3)),'red','white'))
#   superheat(DE %>% filter(cluster==cluster_num) %>% select(1:36), 
#             title = paste0('Cluster ',i),
#             pretty.order.rows = T, 
#             pretty.order.cols = F, 
#             bottom.label.text.angle  = 90,
#             bottom.label.text.size = 3,
#             bottom.label.size = 0.7,
#             left.label = 'none',
#             #heat.pal = c("#386cb0","#7fc97f","white"),
#             yt = comparisons_to_highlight$Size,
#             yt.obs.col = comparisons_to_highlight$Points,
#             yt.num.ticks = 0, yt.axis.name = '', yt.plot.size = 0.05)
# }
# # get correlations of each comparison
# DE %>% filter(cluster==3) %>% select(1:36) %>% summarise_each(funs(var))
```