---
title: "DE"
output:
  html_document: default
  html_notebook: default
---

We begin by loading in the the raw data, excluding files which fail QC due to either 

```{r, message = FALSE}
library(tximport)
library(limma)
library(edgeR)
library(qsmooth)
source('~/git/scripts/theme_Publication.R')
source('~/git/unified_gene_expression/scripts/parse_sample_attribute.R')
# low median count files removed, from 'analysis/QC_and_qsmooth.Rmd'
load('~/git/unified_gene_expression/data/lengthScaledTPM_processed_01_27_2017.Rdata')
```

Clean up metadata, removing run accession, and then removing dups. Some samples have multiple runs
```{r, message = FALSE}
core_tight <- core_tight %>% dplyr::select(-run_accession)
core_tight <- core_tight[!duplicated(core_tight),]
core_tight$Tissue = trimws(core_tight$Tissue)
```

Pull in salmon counts data directories
```{r, message = FALSE}
# biowulf Salmon counts 
working_dir <- '/data/mcgaugheyd/projects/nei/mcgaughey/unified_gene_expression/salmon_counts_bootstrap50_txUsed/'
# eyeMac
working_dir <- '/Volumes/ThunderBay/PROJECTS/mcgaughey/unified_gene_expression/salmon_counts_bootstrap50_txUsed/'
setwd(working_dir)
files <- list.files(path=working_dir,recursive=TRUE,pattern='quant.sf')
```

Eye only samples, with outliers removed, from previous analyses
```{r, message = FALSE}
eye_samples <- core_tight %>% 
  filter(sample_accession %in% colnames(lengthScaledTPM_qsmooth_highExp_remove_lowGenes)) %>%  # low median count outliers
  filter(!Tissue %in% 'ENCODE Cell Line') %>% 
  filter(study_accession!='SRP012682') %>% 
  filter(!sample_accession %in% c('SRS523795','SRS360124','SRS360123')) %>%  # retina in RPE
  .[['sample_accession']] 
```

Now remove outlier files and ENCODE files. Then select representative subset of GTEx by matching number of tissues to eye tissues and taking equal representation across the 22 GTEx tissues
```{r, message = FALSE} 
set.seed(123421)
sample_number <- length(eye_samples) /
  (core_tight %>% filter(sample_accession %in% colnames(lengthScaledTPM_qsmooth_highExp_remove_lowGenes)) %>% filter(study_accession=='SRP012682') %>% dplyr::select(Tissue) %>% unique() %>% nrow())
sample_number <- round(sample_number)
gtex_sub_samples <- core_tight %>% 
  filter(sample_accession %in% colnames(lengthScaledTPM_qsmooth_highExp_remove_lowGenes)) %>% # low median count outliers
  filter(!Tissue %in% 'ENCODE Cell Line') %>% 
  filter(!sample_accession %in% eye_samples) %>% 
  filter(!sample_accession %in% c('SRS523795','SRS360124','SRS360123')) %>%  # retina in RPE
  filter(!sample_accession %in% c('SRS332999', 'SRS389531', 'SRS626623', 'SRS627133', 'SRS623923', 'SRS629483')) %>% # few gtex samples that outliers (in weird clusters)
  group_by(Tissue) %>% 
  sample_n(sample_number) %>% 
  .[['sample_accession']]
eye_and_gtex_samples <- c(eye_samples, gtex_sub_samples)
```

Update files to import
```{r, message = FALSE}
eye_and_gtex_files <- files[files %in% paste(eye_and_gtex_samples,'/quant.sf',sep='')]
eye_files <- files[files %in% paste(eye_samples,'/quant.sf',sep='')]
```
Gene TX to name conversion annotations
```{r, message = FALSE}
load('~/git/unified_gene_expression/data/gencode_v25_annotation.Rdata')
anno <- gencode_v25_annotation %>% dplyr::select(Transcript.ID, Gene.Name)
```

Create metadata file for eye samples
```{r, message = FALSE}
# tximport to aggregate counts to gene level
#txi.eye_and_gtex <- tximport(eye_and_gtex_files, type = "salmon", tx2gene = anno, reader = read_tsv)
#txi.eye <- tximport(eye_files, type = "salmon", tx2gene = anno, reader = read_tsv)

# metadata
SampleTable_eye <- data.frame(eye_samples) %>% 
  mutate(sample_accession = eye_samples) %>% 
  select(sample_accession) %>% 
  left_join(.,core_tight) %>% 
  mutate(Origin = ifelse(grepl('htert',sample_attribute,ignore.case = T),'hTERT Cell Line',Origin)) %>% 
  select(sample_accession, Tissue, Origin)
```
DESeq2
```{r, message = FALSE}
#############
#DESeq TOO DAMN SLOW
# DESeq object creation for DE
# dds_eye <- DESeqDataSetFromTximport(txi.eye, SampleTable, ~Tissue + Origin)

# differential expression analysis
#DESeq2Table <- DESeq(dds_eye)
#DESeq2Res <- results(DESeq2Table, filterFun=ihw, addMLE=TRUE)
#print(head(DESeq2Res))
#print(table(DESeq2Res$padj < 0.05))
###############
```

tximport for limma voom
```{r, message = FALSE}
txi.eye.lsTPM_full <- tximport(eye_files, type = "salmon", tx2gene = anno, reader = read_tsv, countsFromAbundance = "lengthScaledTPM")
txi.eye.lsTPM <- data.frame(txi.eye.lsTPM_full$counts)
colnames(txi.eye.lsTPM) <- eye_sample
```

Function to do some QC steps and make density plots before and after doing:
1. Remove low count genes
2. qsmooth
Returns qsmooth'ed set with low count genes removed
```{r, message = FALSE}

QC <- function(data, metadata, qsmoothFactor){
  # density plot, pre low gene count removal and qsmooth(?)
  gather_lST<-gather(data, sample_accession) %>% left_join(.,metadata)
  print(ggplot(gather_lST,aes(x=log2(value+1),group=sample_accession)) +
          geom_density() +
          facet_wrap(~Tissue) + 
          theme_Publication())

  # remove genes meeting one of two conditions:
  table(rowSums(data)<(ncol(data)))
  data_geneRemoval <- data[(rowSums(data)>ncol(data)),]

  # re-do density plot
  gather_lST<-gather(data_geneRemoval,sample_accession) %>% left_join(.,metadata)
  print(ggplot(gather_lST,aes(x=log2(value+1),group=sample_accession)) +
    geom_density() +
    facet_wrap(~Tissue) +
    theme_Publication())

  # qsmooth
  qs <- qsmooth(object = data_geneRemoval,groupFactor = as.factor(metadata[,qsmoothFactor]))
  lsTPM_qsmooth <- qsmoothData(qs)

  # re-redo density plot
  gather_lST<-gather(data.table(lsTPM_qsmooth),sample_accession) %>% left_join(.,metadata)
  print(ggplot(gather_lST,aes(x=log2(value+1),group=sample_accession)) +
          geom_density() +
          facet_wrap(~Tissue) +
          theme_Publication())
  return(lsTPM_qsmooth)
}
```

QC eye samples
```{r, message = FALSE}
lsTPM_qsmooth <- QC(txi.eye.lsTPM, SampleTable_eye, 'Tissue')
```


Set up design matrix and do limma voom DE
```{r, message = FALSE}
# design matrix
tissue_origin <- paste(SampleTable_eye$Tissue, SampleTable_eye$Origin,sep="_") %>% str_replace_all(., " ", ".") %>% factor()
design <- model.matrix(~0 + tissue_origin)
colnames(design) <- levels(tissue_origin)
# limma voom
y <- DGEList(lsTPM_qsmooth)
y <- calcNormFactors(y)
v <- voom(y, design)

# set up comparisons
cont.matrix <- makeContrasts(RPE.adult_vs_RPE.cell="RPE_Adult.Tissue-RPE_Cell.Line",
                             RPE.adult_vs_RPE.fetal="RPE_Adult.Tissue-RPE_Fetal.Tissue",
                             RPE.adult_vs_RPE.hTERT="RPE_Adult.Tissue-RPE_hTERT.Cell.Line",
                             RPE.cell_vs_RPE.hTERT="RPE_Cell.Line-RPE_hTERT.Cell.Line",
                             RPE.cell_vs_RPE.fetal="RPE_Cell.Line-RPE_Fetal.Tissue",
                             RPE.adult_vs_Retina.adult="RPE_Adult.Tissue-Retina_Adult.Tissue",
                             RPE.cell_vs_Retina.cell="RPE_Cell.Line-Retina_Cell.Line",
                             RPE.cell_vs_ESC.cell="RPE_Cell.Line-ESC_Cell.Line",
                             Retina.adult_vs_Retina.cell="Retina_Adult.Tissue-Retina_Cell.Line",
                             Retina.cell_vs_ESC.cell="Retina_Cell.Line-ESC_Cell.Line",
                             Cornea.adult_vs_RPE.adult="Cornea_Adult.Tissue-RPE_Adult.Tissue",
                             Cornea.adult_vs_Retina.adult="Cornea_Adult.Tissue-Retina_Adult.Tissue",
                            #RPE_vs_Retina="(RPE_Adult.Tissue+RPE_Cell.Line)-(Retina_Adult.Tissue+Retina_Cell.Line)",
                             levels=design)

# fitting time
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=cont.matrix)
efit <- eBayes(vfit)
# summary, with log FC > 1 and adj.p<0.05
summary(decideTests(efit, lfc = 1, p.value = 0.05))
```

Extract stats for comparisons
```{r, message = FALSE}
topTable(efit,coef=2,adjust="fdr",number = 30000, p.value = 0.05, confint = TRUE) %>% rownames_to_column('Gene') %>% arrange(abs(logFC))
```


