---
title: "DE"
output:
  html_document: default
  html_notebook: default
---

We begin by loading in the the raw data, excluding files which fail QC due to either 

```{r, message = FALSE}
library(tximport)
library(limma)
library(edgeR)

source('~/git/unified_gene_expression/scripts/parse_sample_attribute.R')
# low median count files removed, from 'analysis/QC_and_qsmooth.Rmd'
load('~/git/unified_gene_expression/data/lengthScaledTPM_processed_01_27_2017.Rdata')


# clean up metadata, removing run accession, and then removing dups. Some samples have multiple runs
core_tight <- core_tight %>% dplyr::select(-run_accession)
core_tight <- core_tight[!duplicated(core_tight),]
core_tight$Tissue = trimws(core_tight$Tissue)

# biowulf Salmon counts 
working_dir <- '/data/mcgaugheyd/projects/nei/mcgaughey/unified_gene_expression/salmon_counts_bootstrap50_txUsed/'
# eyeMac
working_dir <- '/Volumes/ThunderBay/PROJECTS/mcgaughey/unified_gene_expression/salmon_counts_bootstrap50_txUsed/'
setwd(working_dir)
files <- list.files(path=working_dir,recursive=TRUE,pattern='quant.sf')

# eye only samples, with outliers removed
eye_samples <- core_tight %>% 
  filter(sample_accession %in% colnames(lengthScaledTPM_qsmooth_highExp_remove_lowGenes)) %>%  # low median count outliers
  filter(!Tissue %in% 'ENCODE Cell Line') %>% 
  filter(study_accession!='SRP012682') %>% 
  filter(!sample_accession %in% c('SRS523795','SRS360124','SRS360123')) %>%  # retina in RPE
  .[['sample_accession']] 

# remove outlier files
# and ENCODE
# select representative subset of GTEx by matching number of tissues to eye tissues
# and taking equal representation across the 22 GTEx tissues
# there are 
set.seed(123421)
sample_number <- length(eye_samples) /
  (core_tight %>% filter(sample_accession %in% colnames(lengthScaledTPM_qsmooth_highExp_remove_lowGenes)) %>% filter(study_accession=='SRP012682') %>% dplyr::select(Tissue) %>% unique() %>% nrow())
sample_number <- round(sample_number)
gtex_sub_samples <- core_tight %>% 
  filter(sample_accession %in% colnames(lengthScaledTPM_qsmooth_highExp_remove_lowGenes)) %>% # low median count outliers
  filter(!Tissue %in% 'ENCODE Cell Line') %>% 
  filter(!sample_accession %in% eye_samples) %>% 
  filter(!sample_accession %in% c('SRS523795','SRS360124','SRS360123')) %>%  # retina in RPE
  filter(!sample_accession %in% c('SRS332999', 'SRS389531', 'SRS626623', 'SRS627133', 'SRS623923', 'SRS629483')) %>% # few gtex samples that outliers (in weird clusters)
  group_by(Tissue) %>% 
  sample_n(sample_number) %>% 
  .[['sample_accession']]
eye_and_gtex_samples <- c(eye_samples, gtex_sub_samples)

# update files to import
eye_and_gtex_files <- files[files %in% paste(eye_and_gtex_samples,'/quant.sf',sep='')]
eye_files <- files[files %in% paste(eye_samples,'/quant.sf',sep='')]

# Gene TX to name conversion
load('~/git/unified_gene_expression/data/gencode_v25_annotation.Rdata')
anno <- gencode_v25_annotation %>% dplyr::select(Transcript.ID, Gene.Name)

# tximport to aggregate counts to gene level
#txi.eye_and_gtex <- tximport(eye_and_gtex_files, type = "salmon", tx2gene = anno, reader = read_tsv)
#txi.eye <- tximport(eye_files, type = "salmon", tx2gene = anno, reader = read_tsv)

# metadata
SampleTable_eye <- data.frame(eye_samples) %>% 
  mutate(sample_accession = eye_samples) %>% 
  select(sample_accession) %>% 
  left_join(.,core_tight) %>% 
  select(sample_accession, Tissue, Origin)


#############
#DESeq TOO DAMN SLOW
# DESeq object creation for DE
# dds_eye <- DESeqDataSetFromTximport(txi.eye, SampleTable, ~Tissue + Origin)

# differential expression analysis
#DESeq2Table <- DESeq(dds_eye)
#DESeq2Res <- results(DESeq2Table, filterFun=ihw, addMLE=TRUE)
#print(head(DESeq2Res))
#print(table(DESeq2Res$padj < 0.05))
###############

# length scaled TPM for limma voom
txi.eye.lsTPM <- tximport(eye_files, type = "salmon", tx2gene = anno, reader = read_tsv, countsFromAbundance = "lengthScaledTPM")

# design matrix
tissue_origin <- paste(SampleTable_eye$Tissue, SampleTable_eye$Origin,sep="_") %>% str_replace_all(., " ", ".") %>% factor()
design <- model.matrix(~0 + tissue_origin)
colnames(design) <- levels(tissue_origin)
# limma voom
y <- DGEList(txi.eye.lsTPM$counts)
y <- calcNormFactors(y)
v <- voom(y, design)
fit <- lmFit(v)
fit <- eBayes(fit)
# set up comparisons
cont.matrix <- makeContrasts(RPE.adult_vs_Retina.adult="RPE_Adult.Tissue-Retina_Adult.Tissue",
                             RPE.cell_vs_Retina.cell="RPE_Cell.Line-Retina_Cell.Line",
                             RPE.cell_vs_ESC.cell="RPE_Cell.Line-ESC_Cell.Line",
                             Retina.cell_vs_ESC.cell="Retina_Cell.Line-ESC_Cell.Line",
                             Cornea.adult_vs_RPE.adult="Cornea_Adult.Tissue-RPE_Adult.Tissue",
                             Cornea.adult_vs_Retina.adult="Cornea_Adult.Tissue-Retina_Adult.Tissue",
                             levels=design)
fit2  <- contrasts.fit(fit, cont.matrix)
fit2  <- eBayes(fit2)
topTable(fit2,coef=2,adjust="fdr",number = 100000, p.value = 0.01, confint = TRUE) %>% rownames_to_column('Gene') %>% arrange(logFC)
```

```{r, message = FALSE}
ddsGA <- DESeqDataSetFromTximport(txi, SampleTable, ~GenoAge + Non_Exon_Alignment_Fraction)
```


We first create a __DESeqDataSet__ object. We then perform the differential expression analysis (Retina vs. RPE), first with no batch correction. We find 434 significant genes ($p<0.05$).

```{r, message = FALSE}
dds <- DESeqDataSetFromMatrix(countData = expr,
                              colData = meta,
                              design = ~ Tissue)
DESeq2Table <- DESeq(dds)
DESeq2Res <- results(DESeq2Table, filterFun=ihw, addMLE=TRUE)
print(head(DESeq2Res))
print(table(DESeq2Res$padj < 0.05))

# hist of pvalues
hist(DESeq2Res$pvalue, col = "lavender", main = "Retina vs RPE, no correction", xlab = "p-values")

# Filter by padj and padjust to get rid of NA
DESeq2Res <- DESeq2Res[ !is.na(DESeq2Res$padj), ]
DESeq2Res <- DESeq2Res[ !is.na(DESeq2Res$pvalue), ]

# fdr ztool to correct pvalues
DESeq2Res <- DESeq2Res[, -which(names(DESeq2Res) == "padj")]
FDR.DESeq2Res <- fdrtool(DESeq2Res$stat, statistic= "normal", plot = F)
DESeq2Res[,"padj"]  <- adj_pvalues(ihw(FDR.DESeq2Res$pval, DESeq2Res$baseMean, 0.05))
hist(FDR.DESeq2Res$pval, col = "royalblue4", 
     main = "Retina vs RPE, corrected null model", xlab = "CORRECTED p-values")

# MA plot
DESeq2::plotMA(DESeq2Res)

# output data
stats <- data.frame(DESeq2Res)
gene_counts <- counts(DESeq2Table, normalized=TRUE)

# merge together
out <- merge(stats,gene_counts,by="row.names")
colnames(out)[1] <- 'Gene'

# order by pvalue and add FC column
out <- out %>% arrange(pvalue) %>% mutate(FoldChange = 2^log2FoldChange) %>% dplyr::dplyr::select(Gene:padj,FoldChange)

out.sig <- filter(out, padj < 0.05)

write.csv(out, file = "~/Documents/git/unified_gene_expression/results/DE_full_noCorrection.csv")
write.csv(out.sig, file = "~/Documents/git/unified_gene_expression/results/DE_pAdj0.05_noCorrection.csv")

write_delim(as.data.frame(out.sig$Gene), delim = "\n", path = "~/Documents/git/unified_gene_expression/results/DE_pAdj0.05_noCorrection_geneList.txt")
```

# Batch Issue Investigation

We now consider investigating possible batch issues. The two possible batch effects we explore are *study_accession* and whether the sample comes from adult tissue or fetal tissue (all of the fetal tissue is from cell lines).

```{r, message = FALSE}
set.seed(47)

perplexities = seq(10, 40, by = 10)

for(p in perplexities){
  
  tsne_out = Rtsne(t(expr), perplexity = p, check_duplicates = F, theta = 0.0)
  tsne_plot = data.frame(tsne_out$Y)
  
  p1 = tsne_plot %>%
    ggplot(aes(x=X1,y=X2,colour=factor(meta$study_accession))) + 
    geom_point(size=4) + ggtitle(paste0("t-sne. Perplexity = ", p))
  print(p1)
  
  p2 = tsne_plot %>%
    ggplot(aes(x=X1,y=X2,colour=factor(meta$Origin))) + 
    geom_point(size=4) + ggtitle(paste0("t-sne. Perplexity = ", p))
  print(p2)
}
```
The clearest evidence of clustering occurs with a perplexity of 20. The clusters are clearly defined by the variable *study_accession*. As such, we consider incorporating *study_accession* into our differential expression model. We perform the differential expression analysis (Retina vs. RPE), first with no batch correction. We find 443 significant genes ($p<0.05$).

```{r, message = FALSE}
dds <- DESeqDataSetFromMatrix(countData = expr,
                              colData = meta,
                              design = ~ Tissue + study_accession)
DESeq2Table <- DESeq(dds)
DESeq2Res <- results(DESeq2Table, filterFun=ihw, contrast = c("Tissue", "Retina", "RPE"), addMLE=TRUE)
print(head(DESeq2Res))
print(table(DESeq2Res$padj < 0.05))

# hist of pvalues
hist(DESeq2Res$pvalue, col = "lavender", main = "Retina vs RPE, no correction", xlab = "p-values")

# Filter by padj and padjust to get rid of NA
DESeq2Res <- DESeq2Res[ !is.na(DESeq2Res$padj), ]
DESeq2Res <- DESeq2Res[ !is.na(DESeq2Res$pvalue), ]

# fdr ztool to correct pvalues
DESeq2Res <- DESeq2Res[, -which(names(DESeq2Res) == "padj")]
FDR.DESeq2Res <- fdrtool(DESeq2Res$stat, statistic= "normal", plot = F)
DESeq2Res[,"padj"]  <- adj_pvalues(ihw(FDR.DESeq2Res$pval, DESeq2Res$baseMean, 0.05))
hist(FDR.DESeq2Res$pval, col = "royalblue4", 
     main = "Retina vs RPE, corrected null model", xlab = "CORRECTED p-values")

# MA plot
DESeq2::plotMA(DESeq2Res)

# output data
stats <- data.frame(DESeq2Res)
gene_counts <- counts(DESeq2Table, normalized=TRUE)

# merge together
out.study.correct <- merge(stats,gene_counts,by="row.names")
colnames(out.study.correct)[1] <- 'Gene'

# order by pvalue and add FC column
out.study.correct <- out.study.correct %>% arrange(pvalue) %>% mutate(FoldChange = 2^log2FoldChange) %>% dplyr::dplyr::select(Gene:padj,FoldChange)

out.sig.study.correct <- filter(out.study.correct, padj < 0.05)

write.csv(out.study.correct, file = "~/Documents/git/unified_gene_expression/results/DE_full_correction_studyAccession.csv")
write.csv(out.sig.study.correct, file = "~/Documents/git/unified_gene_expression/results/DE_pAdj0.05_correction_studyAccession.csv")

write_delim(as.data.frame(out.sig.study.correct$Gene), delim = "\n", path = "~/Documents/git/unified_gene_expression/results/DE_pAdj0.05_correction_studyAccession_geneList.txt")
```

Between the 434 genes found significant with no incorporation of *study_accession* and the 443 genes found significant with *study_accession* incorporated, there was an intersection of 215 genes.

```{r, message = FALSE}
length(intersect(out.sig$Gene, out.sig.study.correct$Gene))
```

# GO term annotation analysis
We conduct GO term annotation using the PANTHER classification system (http://pantherdb.org/tools/compareToRefList.jsp), by GeneOntology. This is performed through the corresponding web app.
